File: compiler/src/main/java/com/aaroncoplan/waterfall/compiler/statements/helpers/Translatable.java
package com.aaroncoplan.waterfall.compiler.statements.helpers;

import com.aaroncoplan.waterfall.compiler.symboltables.SymbolTable;
import com.aaroncoplan.waterfall.compiler.target.CodeGenerator;

public interface Translatable {
    VerificationResult verify(SymbolTable symbolTable);

    /**
     * Render this node via the given backend. Implementations should dispatch to the
     * matching emit* method on {@link CodeGenerator} (e.g. {@code g.emitIfBlock(this)}).
     */
    String translate(CodeGenerator backend);
}

File: compiler/src/main/kotlin/com/aaroncoplan/waterfall/compiler/statements/helpers/Translatable.kt
package com.aaroncoplan.waterfall.compiler.statements.helpers

import com.aaroncoplan.waterfall.compiler.target.CodeGenerator

interface Translatable {

    /**
     * Render this node via the given backend. Implementations should dispatch to the
     * matching `emit*` method on [CodeGenerator] (e.g. `g.emitIfBlock(this)`).
     *
     * Note: `verify(symbolTable)` has been removed from this interface in §5.3.
     * Verification now lives in `compiler/.../verifier/` — see `Verifier.verifyModule`.
     */
    fun translate(backend: CodeGenerator): String
}
